dev/PushNotifications/PushManager.cpp

```diff
@@ -4,11 +4,111 @@
+
+// Error code - platform reconnecting with wns
+#define WNP_E_RECONNECTING  0x880403E9L
+
+// Error code - Current bind operation failed because another bind operation in progress
+#define WNP_E_BIND_USER_BUSY 0x880403FEL
 
 namespace winrt::Microsoft::ProjectReunion::implementation
 {
+    std::list<winrt::guid> PushManager::remoteIdList;
+    std::mutex PushManager::mtx;
+    std::unique_lock<std::mutex> PushManager::lock (mtx, std::defer_lock); // Avoid locking during the constructor call
+
+    bool PushManager::isChannelRequestRetryable(const winrt::hresult& hrException)
+    {
+        switch (hrException)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+            // E_TIMEOUT
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING
+        case WNP_E_BIND_USER_BUSY:
+            return true;
+        default:
+            return false;
+        }
+    }
+
     Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::ChannelResult, Microsoft::ProjectReunion::ChannelResult> PushManager::CreateChannelAsync(winrt::guid remoteId)
     {
-        throw hresult_not_implemented();
+        winrt::Microsoft::ProjectReunion::ChannelResult channelResult{ nullptr };
+        auto progress{ co_await winrt::get_progress_token() };
+
+       GUID remoteIdAbiGuid = reinterpret_cast<GUID&>(remoteId);
+       if (remoteIdAbiGuid == GUID_NULL)
+        {
+            channelResult = winrt::make<winrt::Microsoft::ProjectReunion::implementation::ChannelResult>(nullptr, E_INVALIDARG, ChannelStatus::CompletedFailure);
+            co_return channelResult;
+        }
+
+       lock.lock();
+       if (std::find(remoteIdList.begin(), remoteIdList.end(), remoteId) != remoteIdList.end())
+       {
+           progress(winrt::make<winrt::Microsoft::ProjectReunion::implementation::ChannelResult>(nullptr, WPN_E_OUTSTANDING_CHANNEL_REQUEST, ChannelStatus::InProgress));
+           channelResult = winrt::make<winrt::Microsoft::ProjectReunion::implementation::ChannelResult>(nullptr, WPN_E_OUTSTANDING_CHANNEL_REQUEST, ChannelStatus::CompletedFailure);
+           co_return channelResult;
+       }
+       else
+       {
+           remoteIdList.push_back(remoteId);
+       }
+       lock.unlock();
+
+       PushNotificationChannelManager channelManager{};
+
+       for (auto backOffTimeInSeconds = 30; backOffTimeInSeconds <= MAX_BACKOFF_SECONDS * 2; backOffTimeInSeconds *= 2)
+       {
+           try
+           {
+               auto pushChannel = co_await channelManager.CreatePushNotificationChannelForApplicationAsync();
+
+               if (pushChannel != nullptr)
+               {
+                   channelResult = winrt::make<winrt::Microsoft::ProjectReunion::implementation::ChannelResult>(pushChannel, S_OK, ChannelStatus::CompletedSuccess);
+                   break;
+               }
+           }
+           catch (winrt::hresult_error const& ex)
```

Comment:
This doesn't catch E_OUTOFMEMORY or some other exceptions that convert to std exception types.

@kennykerr what's the recommended practice? Should there also be a additional catch blocks? For what types? Or catch(...) for other cases? Or just one catch(...)? Should the else case here be omitted?

Reply:
It's only `std::bad_alloc` that's excluded. If you want to get back an `HRESULT` that includes `E_OUTOFMEMORY` (and all `hresult` and `std` exceptions) you can do something like this:

```C++
catch (...)
{
    auto ex = hresult_error(to_hresult(), take_ownership_from_abi);
    // ex.code() now includes E_OUTOFMEMORY
}
```

This is preferred to just using `auto hr = to_hresult();` because the latter leaves any error information on the thread that can confuse subsequent code.